gossip/hub/handlers.go

```diff
@@ -0,0 +1,664 @@
+		return http.StatusInternalServerError, fmt.Errorf("failed to marshal tree head: %v", err)
+	}
+	signature, err := c.signer.Sign(hthData)
+	if err != nil {
+		return http.StatusInternalServerError, fmt.Errorf("failed to sign STH data: %v", err)
+	}
+
+	// Now build the final result object and marshal to JSON
+	jsonRsp := api.GetSTHResponse{TreeHeadData: hthData, HubSignature: signature}
+	jsonData, err := json.Marshal(&jsonRsp)
+	if err != nil {
+		return http.StatusInternalServerError, fmt.Errorf("failed to marshal response: %v %v", jsonRsp, err)
+	}
+
+	w.Header().Set(contentTypeHeader, contentTypeJSON)
+	if _, err := w.Write(jsonData); err != nil {
+		return http.StatusInternalServerError, fmt.Errorf("failed to write response data: %v", err)
+	}
+
+	return http.StatusOK, nil
+}
+
+func getSTHConsistency(ctx context.Context, c *hubInfo, w http.ResponseWriter, r *http.Request) (int, error) {
+	firstVal := r.FormValue(api.GetSTHConsistencyFirst)
+	if firstVal == "" {
+		return http.StatusBadRequest, errors.New("parameter 'first' is required")
+	}
+	secondVal := r.FormValue(api.GetSTHConsistencySecond)
+	if secondVal == "" {
+		return http.StatusBadRequest, errors.New("parameter 'second' is required")
+	}
+
+	first, err := strconv.ParseInt(firstVal, 10, 64)
+	if err != nil {
+		return http.StatusBadRequest, errors.New("parameter 'first' is malformed")
+	}
+	second, err := strconv.ParseInt(secondVal, 10, 64)
+	if err != nil {
+		return http.StatusBadRequest, errors.New("parameter 'second' is malformed")
+	}
+
+	if first < 0 || second < 0 {
+		return http.StatusBadRequest, fmt.Errorf("first and second params cannot be <0: %d %d", first, second)
+	}
+	if second < first {
+		return http.StatusBadRequest, fmt.Errorf("invalid first, second params: %d %d", first, second)
+	}
+
+	var jsonRsp api.GetSTHConsistencyResponse
+	if first != 0 {
+		req := trillian.GetConsistencyProofRequest{LogId: c.logID, FirstTreeSize: first, SecondTreeSize: second}
+
+		glog.V(2).Infof("%s: GetSTHConsistency(%d, %d) => grpc.GetConsistencyProof %+v", c.hubPrefix, first, second, req)
+		rsp, err := c.rpcClient.GetConsistencyProof(ctx, &req)
+		glog.V(2).Infof("%s: GetSTHConsistency <= grpc.GetConsistencyProof err=%v", c.hubPrefix, err)
+		if err != nil {
+			return c.toHTTPStatus(err), fmt.Errorf("backend GetConsistencyProof request failed: %v", err)
+		}
+
+		// TODO(drysdale): configure the hub with the Trillian public key for this tree, and check the Trillian signature here.
+		var root types.LogRootV1
+		if err := root.UnmarshalBinary(rsp.SignedLogRoot.LogRoot); err != nil {
+			return http.StatusInternalServerError, fmt.Errorf("failed to unmarshal log root: %v", err)
+		}
+
+		// We can get here with a tree size too small to satisfy the proof.
+		if int64(root.TreeSize) < second {
+			return http.StatusBadRequest, fmt.Errorf("need tree size: %d for proof but only got: %d", second, root.TreeSize)
+		}
+
+		if err := checkHashSizes(rsp.Proof.Hashes); err != nil {
```

Comment:
should `checkHashSizes` be part of the `LogClient`?  Seems like that would be a useful thing to have there as well.

Reply:
Maybe, but the `LogClient` verifies proof chains so would automatically fail on mis-sized hashes anyway? (In the personality this is just a sanity check as the data is passing through.)